_posts/2023-06-16-End-to-end-acknowledgements-in-Data-Prepper.md

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title:  "End-to-end acknowledgements in Data Prepper"
+authors:
+- kkondaka
+- dlv
+date: 2023-06-16 10:00:00 -0700
+categories:
+  - technical-post
+
+meta_keywords: Data Prepper, end-to-end acknowledgements
+meta_description: Learn how Data Prepper end-to-end acknowledgements provide data durability by making sure that data is delivered to the sink before the source is notified
+
+---
+
+## The need: Improving durability
+
+
+[Data Prepper](https://opensearch.org/docs/latest/data-prepper/index/) is an open-source data collector. You can use Data Prepper to ingest data into OpenSearch clusters. With Data Prepper, you can filter, enrich, normalize, and transform the data going into your OpenSearch cluster.
+
+Data Prepper comes with an in-memory buffer which allows for fast processing of large volumes of data. Besides throughput, another quality that users often seek in Data Prepper is data durability. They want to ensure that data received by Data Prepper reaches the desired sink.
+
+Several challenges arise with data durability. For example, Data Prepper may exhaust memory or other system resources, or the hardware running Data Prepper may fail. The maintainers of Data Prepper and their teams have observed that the most common hindrance to data durability is the sink. If the OpenSearch cluster is unable to receive data either due to temporary stress on the cluster or if it is under-scaled, then Data Prepper cannot send data to the destination sink.
+
+We have also considered the sources of data. In particular, we have noticed that the Amazon S3 source is already reading data from a highly durable store. Keeping these observations in mind we realized that if Data Prepper fails to deliver data to OpenSearch, it can retry reading from S3. We just need to know when the data is written before deleting the SQS message that notifies Data Prepper of an available S3 object to consume.
+
+## Our solution: End-to-end acknowledgements
+
+Data Prepper provides data durability through the use of end-to-end acknowledgements. When a Data Prepper source is configured to use end-to-end acknowledgements, the source is notified only when the data is successfully delivered to the sink. If the source receives end-to-end acknowledgments, it can take appropriate actions such as removing SQS messages that have been successfully delivered to the sink or incrementing the commit offset in case of Kafka source. If end-to-end acknowledgements are not received, the source may retry operation or notify the external source of the failure. The following picture illustrates the control flow when a Data Prepper source is configured to use end-to-end acknowledgements.
```

Comment:
```suggestion
When a Data Prepper source is configured to use end-to-end acknowledgements, the source is notified only when the data is successfully delivered to the sink. If the source receives end-to-end acknowledgments, it can take appropriate actions, such as removing SQS messages that have been successfully delivered to the sink or incrementing the commit offset in the case of a Kafka source. If end-to-end acknowledgements are not received, the source may retry the operation or notify the external source of the failure. The following diagram shows the control flow of a Data Prepper source configured to use end-to-end acknowledgements.
```